File: frontend/src/services/inspections.ts
import { api, unwrapList } from './api';
import type { Id, Inspection, InspectionPayload } from '@/types';

export const inspectionsService = {
  async list(projectId?: Id): Promise<Inspection[]> {
    const { data } = await api.get('/inspections', { params: projectId ? { project_id: projectId } : undefined });
    return unwrapList<Inspection>(data);
  },
  async create(payload: InspectionPayload): Promise<Inspection> {
    const { data } = await api.post<Inspection>('/inspections', payload);
    return data;
  },
};

File: backend/app/api/v1/inspections.py
from fastapi import APIRouter, Depends, status

from app.api.deps import CurrentUser, DbSession, require_roles
from app.models.enums import Role
from app.schemas.inspection import InspectionCreate, InspectionOut, InspectionUpdate
from app.services import inspection_service

router = APIRouter(prefix="/inspections", tags=["Inspections"])


@router.get("", response_model=list[InspectionOut])
async def list_inspections(db: DbSession, current_user: CurrentUser, project_id: str | None = None) -> list[InspectionOut]:
    return await inspection_service.list_inspections(db, project_id)


@router.post(
    "",
    response_model=InspectionOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_roles(Role.ENGINEER, Role.ADMIN))],
)
async def create_inspection(payload: InspectionCreate, db: DbSession, current_user: CurrentUser) -> InspectionOut:
    return await inspection_service.create_inspection(db, payload, current_user)


@router.put(
    "/{inspection_id}",
    response_model=InspectionOut,
    dependencies=[Depends(require_roles(Role.ENGINEER, Role.ADMIN))],
)
async def update_inspection(inspection_id: str, payload: InspectionUpdate, db: DbSession) -> InspectionOut:
    return await inspection_service.update_inspection(db, inspection_id, payload)
